### DISCUSSION TITLE
Non-remix exports break Vite's HMR

### DISCUSSION THREAD

**Discussion by andrecasal** [epicweb-dev/epic-stack#676] [Q&A]
Vite's HMR only works when a file exports a component and nothing else. Because Remix knows which exports belong to Remix (handle, links, loader, action, etc), it allows us to still export these things from within a route.

However, the `root.tsx` contains an `export function ThemeSwitch()`, which isn't a standard Remix export.

Won't this break HMR? According to [Remix's HMR docs](https://remix.run/docs/en/main/discussion/hot-module-replacement), it should.

After upgrading to Vite, I lost Live Reload/HMR/HDR. I'm not sure exports like these are the culprits, but wanted to ask before I go digging around. Manual refresh still works.

**Comment by andrecasal** [comment]
Just ran `npx create-epic-app@latest` and HMR is broken.

```
[vite] hmr invalidate /app/root.tsx Could not Fast Refresh. Learn more at https://github.com/vitejs/vite-plugin-react/tree/main/packages/plugin-react#consistent-components-exports`
```

**Comment by kentcdodds** [comment]
> However, the root.tsx contains an export function ThemeSwitch(), which isn't a standard Remix export.

This is not true: https://github.com/epicweb-dev/epic-stack/blob/c047d75db53dfca4d73f22b3de8bbb593ec01732/app/root.tsx#L379 🤔


That said, it does export `useTheme` and `useOptimisticThemeMode`. This could be reason enough to move these things out of `root.tsx`. Could you try that and let us know whether that fixes the issue?

**Comment by andrecasal** [comment]
I tried it like this and HMR still doesn't work.

![Screenshot 2024-03-26 at 14 53 15](https://github.com/epicweb-dev/epic-stack/assets/22457463/c641fb13-79c4-4a72-9d27-72afcae27caa)

**Reply by andrecasal** [reply]
Vite seems to be doing it's thing, but the browser doesn't reload. Only a manual reload would update its contents.

![Screenshot 2024-03-26 at 14 54 11](https://github.com/epicweb-dev/epic-stack/assets/22457463/f1a9b64f-62db-45cb-8ef6-3b0093a1c701)

**Reply by kentcdodds** [reply]
Sounds like you need to open an issue on Remix then.

**Reply by andrecasal** [reply]
A barebones Remix installation has HMR working perfectly, so I imagine it has to do with the Epic Stack.

**Reply by kentcdodds** [reply]
Interesting. ~~Does the Indie Stack have working HMR?~~ I just realized the Indie Stack hasn't been updated to the latest Remix which is certainly a surprise. Does the express template work with HMR? I'm just wondering if it's an issue with a custom server: https://github.com/remix-run/remix/tree/main/templates/express

**Comment by thefrana** [comment]
@andrecasal have you resolved this issue? I also encountered it and the problem was in Express integration with Vite. I do not use Epic Stack directly so I am not sure if the problem is the same.
